helix-term/src/ui/popup.rs

```diff
@@ -228,6 +238,41 @@ impl<T: Component> Component for Popup<T> {
 
         let inner = area.inner(&self.margin);
         self.contents.render(inner, surface, cx);
+
+        // render scrollbar if contents do not fit
+        if self.has_scrollbar {
+            let win_height = inner.height as usize;
+            let len = self.child_size.1 as usize;
+            let fits = len <= win_height;
+            let theme = &cx.editor.theme;
+            let scroll = self.scroll;
+
+            const fn div_ceil(a: usize, b: usize) -> usize {
+                (a + b - 1) / b
+            }
+
+            let scroll_height = div_ceil(win_height.pow(2), len).min(win_height as usize);
+            let scroll_line = (win_height - scroll_height) * scroll
+                / std::cmp::max(1, len.saturating_sub(win_height));
+
+            let scroll_style = theme.get("ui.menu.scroll");
+            for (i, _) in (scroll..(scroll + win_height).min(len)).enumerate() {
+                let cell = &mut surface[(area.x + area.width - 1, area.y + i as u16)];
+
+                if !fits {
+                    // Draw scroll track
+                    cell.set_symbol("▐"); // right half block
+                    cell.set_fg(scroll_style.bg.unwrap_or(helix_view::theme::Color::Reset));
+                }
+
+                let is_marked = i >= scroll_line && i < scroll_line + scroll_height;
+
+                if !fits && is_marked {
+                    // Draw scroll thumb
+                    cell.set_fg(scroll_style.fg.unwrap_or(helix_view::theme::Color::Reset));
+                }
```

Comment:
Why not move this into `!fits` block?

Reply:
Basically, the sumbitted PR is just a copy of the already implemented menu scrollbar. You are right, though. The code could use some optimisation. I'm on it!